### PR TITLE
34 integrate canopy height model and plot imputation map onramp code

### PR DESCRIPTION
**Pull request by lindsaywiard**
hag_pim.py holds functions for creating a GeoDataFrame of tree plot points with locations informed by a height above ground raster.

Also included in this branch is a test file which contains pytest functions for testing the functions within hag_pim.py

### CODE REVIEW

File: fastfuels_core/onramps/hag_pim.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        pim_raster,                     
+        pim_raster.rio.nodata              
+    )
+    combined_raster.rio.write_crs(
+        pim_raster.rio.crs, inplace=True)
+    combined_raster.rio.write_nodata(
+        pim_raster.rio.nodata, inplace=True)
+    return combined_raster
+
+def create_gdf(raster):
```
Comment (amarcozzi) [inline fastfuels_core/onramps/hag_pim.py:175]: What about renaming this to something more specific? Something like create_plots_gdf_from_resampled_pim? Having a plot_column_name that defaults to "PLOT_ID" would be helpful here also.

File: fastfuels_core/onramps/hag_pim.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    '''
+    Returns: 
+        plot imputation raster with all cells filled with data
+        - values were determined using nearest neighbor
+    '''
+    pim_interp = pim_raster.rio.interpolate_na(
+        method='nearest')
+    pim_interp.rio.write_crs(pim_raster.rio.crs, inplace=True)
+    pim_interp.rio.write_nodata(pim_raster.rio.nodata, inplace=True)
+    return pim_interp
```
Comment (amarcozzi) [inline fastfuels_core/onramps/hag_pim.py:155]: There's some potential here to combine these into a single interpolate_raster function that takes in the interpolation method, but I'd say that addressing this is low priority.

File: tests/onramps/test_hag_pim.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    toy_raster.rio.write_nodata(nodata_value, inplace=True)
+    return toy_raster
+
+@pytest.fixture
+def hag_4326():
+    data = np.array([
+        [0, 0, 0, 1, 1, 0],
+        [0, 0, 0, 0, 0, 1],
+        [0, 1, 1, 1, 1, 1],
+        [0, 1, 0, 1, 1, 1]
```
Comment (amarcozzi) [inline tests/onramps/test_hag_pim.py:51]: It would be good to diversify this data instead of all edge cases. Maybe have some 0s, 1s, and values from 1-60?

**Review by amarcozzi — CHANGES_REQUESTED** [review]

File: fastfuels_core/onramps/hag_pim.py (hunk 2 above)
Comment (lindsaywiard) [reply fastfuels_core/onramps/hag_pim.py:155]: Yeah, I have wondered about that too. Or we could just interpolate both within one function? I mean both of these steps are just using rasterio functions anyways, do we even really need these to be separate functions?

**Comment by lindsaywiard** [comment]
I have made all the changes besides creating one interpolate_raster function which I commented with a question. I have also updated the tests to match. I'm happy to make another commit, but I had some thoughts and figured we might want to discuss them first.

- I think we need to think about testing the nodata value more robustly. I was going to go back to the proof_of_concept.py and play with real data a bit more to think more carefully about how any issues with the nodata value might arise. Do you have thoughts about where that might happen?
- Should we be testing the sample_plots_from_hag overall function in some way?
- It seems like I am not testing the final gdf very much. Do you have any thoughts on additional issues that might arise there?

File: fastfuels_core/onramps/hag_pim.py (hunk 2 above)
Comment (amarcozzi) [reply fastfuels_core/onramps/hag_pim.py]: I'll roll this into the next review.

**Review by amarcozzi — CHANGES_REQUESTED** [review]
Thanks this is coming together and looking good! There's just a few more things to do and we're really close to merge this.

> * I think we need to think about testing the nodata value more robustly. I was going to go back to the proof_of_concept.py and play with real data a bit more to think more carefully about how any issues with the nodata value might arise. Do you have thoughts about where that might happen?

Good point. Do you see any cases where this could cause a problem? We want to maintain the nodata value for the PIM raster since that is the "null plot" in TreeMap, but it's probably worth changing the nodata values to 0 and setting nodata=0 for the HAG raster right from the beginning in sample_plots_from_hag. Is there anything else that I'm missing?

> * Should we be testing the sample_plots_from_hag overall function in some way?

I feel comfortable not testing this because it doesn't itself do any logic and only combines some of the functions in this module. It's more of a helper function. So to the extent that we trust the tests on our other functions, then we should trust that this is also working correctly.

> * It seems like I am not testing the final gdf very much. Do you have any thoughts on additional issues that might arise there?

I can't think of any concerning cases. It's a pretty straightforward function so we should be good here. Maybe a test case that actually makes sure that the nodata value from the PIM raster is successfully written as a plot identifier? 

In addition to those above there are a few other minor changes to make to match the style and formatting of the rest of the repository:

- [ ] Can you get pre-commit set up and run it before pushing? The [contributing guide](https://github.com/silvxlabs/fastfuels-core/blob/main/CONTRIBUTING.md) in the repository goes into how this happens, and that's what this CI/CD action is doing on the PRs.
- [ ] Can you add type hits to the function parameters similar to how you do it in the sample_plots_from_hag function?
- [ ] Can you change docstrings to the numpy format? We're not able to render the documentation using Sphinx or mkdocs without everything being the same format. Your IDE should have a setting to change docstring types, and [here](https://github.com/silvxlabs/fastfuels-core/blob/main/fastfuels_core/trees.py#L92-L123) is an example from the repository of what those docstrings look like.

**Comment by lindsaywiard** [comment]
Looks good! I can work on this stuff. I looked into the nodata issues this morning. 

For the HAG, I think we can just keep it the same as it comes in throughout the whole function and just account for it in the compute_to_cover function with: hag_cover = xr.where((raster > min_hag)&(raster != raster.rio.nodata), 1.0, 0.0). We can also change it to 0 as long as you are confident that all nodata values will come in as zero. For instance, in the proof_of_concept.py, the HAG nodata value is 255, but there are no cells with a value of 255. If there were, just writing the nodata value to be zero doesn't change those 255's to zero. Just leaving the nodata value alone avoids this issue. 

For the PIM, you say that nodata value is the null plot. Do you mean the null value that it is originally? For instance in proof_of_concept.py, the nodata value comes in as np.int32(2147483647). Are you saying that you would like to keep this the same throughout and have that be a value left in the gdf? That would be easy to do.

**Comment by amarcozzi** [comment]
Good point about the HAG nodata. I fully agree.

Yes, that interpretation for the PIM raster nodata value is correct. We want to maintain that 2147483647 "plot ID" through the whole process.

**Comment by lindsaywiard** [comment]
Great! I can make those happen!

**Review by amarcozzi — APPROVED** [review]
